File: iris-did/src/main/java/city/smartb/iris/did/DIDVerificationMethodBuilder.java
package city.smartb.iris.did;

import city.smartb.iris.did.model.ControledJsonLdObject;
import city.smartb.iris.did.model.DIDVerificationMethod;
import city.smartb.iris.jsonld.JsonLdObject;
import city.smartb.iris.jsonld.reader.JsonFieldReader;
import com.nimbusds.jose.jwk.JWK;
import com.nimbusds.jose.jwk.KeyUse;
import com.nimbusds.jose.jwk.RSAKey;

import javax.annotation.Nullable;
import java.security.PublicKey;
import java.security.interfaces.RSAPublicKey;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

public class DIDVerificationMethodBuilder extends JsonLdObject {

	@Nullable
	private PublicKey publicKey;

	public static DIDVerificationMethodBuilder create() {
		return new DIDVerificationMethodBuilder(new LinkedHashMap<>(), null);
	}

	public DIDVerificationMethodBuilder(Map<String, Object> json, PublicKey pubkey)
	{
		super(json);
		this.publicKey = pubkey;
	}

	public DIDVerificationMethodBuilder withId(String id) {
		jsonLdObject.put(JsonLdObject.JSON_LD_ID, id);
		return this;
	}
	
	public DIDVerificationMethodBuilder withController(String controller) {
		jsonLdObject.put(ControledJsonLdObject.JSON_LD_CONTROLER, controller);
		return this;
	}

	public DIDVerificationMethodBuilder withType(String type) {
		jsonLdObject.put(JsonLdObject.JSON_LD_TYPE, type);
		return this;
	}

	public DIDVerificationMethodBuilder withPublicKey(PublicKey key) {
		this.publicKey = key;
		return this;
	}

	public DIDVerificationMethod build() {
		addPublicKeyProperties();
		return new DIDVerificationMethod(this.jsonLdObject);
	}

	private DIDVerificationMethodBuilder addPublicKeyProperties() {
		if (this.publicKey == null) return this;

		if (this.getType() == DIDVerificationMethod.RSA_VERIFICATION_2018) {

			JWK jwk = new RSAKey.Builder((RSAPublicKey)this.publicKey)
					.keyUse(KeyUse.SIGNATURE)
					.keyID(UUID.randomUUID().toString())
					.build();

			jsonLdObject.put(DIDVerificationMethod.JSON_LD_PUBLICKEYJWK, jwk.toPublicJWK().toJSONObject());
		}

		return this;
	}
}

File: iris-ld/iris-did/src/main/kotlin/city/smartb/iris/ld/did/DIDVerificationMethodBuilder.kt
package city.smartb.iris.ld.did

import city.smartb.iris.ld.jsonld.JsonLdObject
import com.nimbusds.jose.jwk.JWK
import com.nimbusds.jose.jwk.KeyUse
import com.nimbusds.jose.jwk.RSAKey
import java.security.PublicKey
import java.security.interfaces.RSAPublicKey
import java.util.UUID

class DIDVerificationMethodBuilder(json: Map<String, Any>, private var publicKey: PublicKey?) :
    JsonLdObject(json) {

    companion object {
        fun create(): DIDVerificationMethodBuilder {
            return DIDVerificationMethodBuilder(LinkedHashMap(), null)
        }
    }

    fun withId(id: String): DIDVerificationMethodBuilder {
        jsonLdObject[JSON_LD_ID] = id
        return this
    }

    fun withController(controller: String): DIDVerificationMethodBuilder {
        jsonLdObject[ControledJsonLdObject.JSON_LD_CONTROLER] = controller
        return this
    }

    fun withType(type: String): DIDVerificationMethodBuilder {
        jsonLdObject[JSON_LD_TYPE] = type
        return this
    }

    fun withPublicKey(key: PublicKey): DIDVerificationMethodBuilder {
        publicKey = key
        return this
    }

    fun build(): DIDVerificationMethod {
        addPublicKeyProperties()
        return DIDVerificationMethod(jsonLdObject)
    }

    private fun addPublicKeyProperties(): DIDVerificationMethodBuilder {
        if (publicKey == null) return this
        if (DIDVerificationMethod.RSA_VERIFICATION_2018 == this.type) {
            val jwk: JWK = RSAKey.Builder(publicKey as RSAPublicKey?)
                .keyUse(KeyUse.SIGNATURE)
                .keyID(UUID.randomUUID().toString())
                .build()
            jsonLdObject[DIDVerificationMethod.JSON_LD_PUBLICKEYJWK] = jwk.toPublicJWK().toJSONObject()
            return this
        }
        return this
    }
}
